src/main/java/book/store/onlinebookstore/service/impl/ShoppingCartServiceImpl.java

```diff
@@ -0,0 +1,92 @@
+
+@Service
+@RequiredArgsConstructor
+public class ShoppingCartServiceImpl implements ShoppingCartService {
+    private final ShoppingCartRepository shoppingCartRepository;
+    private final ShoppingCartMapper shoppingCartMapper;
+    private final CartItemRepository cartItemRepository;
+    private final CartItemMapper cartItemMapper;
+    private final BookRepository bookRepository;
+    private final UserRepository userRepository;
+
+    @Override
+    @Transactional
+    public ShoppingCartDto addCartItem(CreateCartItemRequestDto requestDto, User user) {
+        Optional<CartItem> cartItemByBookId = cartItemRepository
+                .findCartItemByShoppingCartIdAndBookId(user.getId(), requestDto.bookId());
+        if (cartItemByBookId.isPresent()) {
+            CartItem cartItem = cartItemByBookId.get();
+            return updateCartItem(user.getId(), cartItem.getId(), requestDto.quantity());
+        }
+
+        CartItem cartItem = cartItemMapper.toCartItem(requestDto);
+        Book bookFromDb = bookRepository.findById(requestDto.bookId())
+                .orElseThrow(() -> new EntityNotFoundException("Can't find book by id "
+                        + requestDto.bookId()));
+        cartItem.setBook(bookFromDb);
+        ShoppingCart shoppingCart = getOrCreateShoppingCart(user);
+        shoppingCart.addCartItem(cartItem);
+
+        cartItemRepository.save(cartItem);
+        return shoppingCartMapper.toShoppingCartDto(shoppingCart);
+    }
+
+    @Override
+    public ShoppingCartDto getShoppingCart(Long userId) {
+        ShoppingCart shoppingCart = shoppingCartRepository.findById(userId)
+                .orElseThrow(() -> new EntityNotFoundException("The shopping cart doesn’t exist"));
+        return shoppingCartMapper.toShoppingCartDto(shoppingCart);
+    }
+
+    @Override
+    @Transactional
+    public ShoppingCartDto updateCartItem(Long userId, Long itemId, Integer quantity) {
+        CartItem cartItem = cartItemRepository.findByIdAndShoppingCartId(itemId, userId)
+                .orElseThrow(() -> new EntityNotFoundException("Can't find item by id "
+                        + itemId + " in your shopping cart"));
+        cartItem.setQuantity(quantity);
+        return shoppingCartMapper.toShoppingCartDto(cartItem.getShoppingCart());
+    }
+
+    @Override
+    @Transactional
+    public ShoppingCartDto deleteCartItemById(Long userId, Long itemId) {
+        ShoppingCart shoppingCart = shoppingCartRepository.findById(userId)
+                .orElseThrow(() -> new EntityNotFoundException("The shopping cart doesn’t exist"));
+        CartItem cartItem = cartItemRepository.findByIdAndShoppingCartId(itemId, userId)
+                .orElseThrow(() -> new EntityNotFoundException("Can't find item by id "
+                        + itemId + " in your shopping cart"));
+        cartItemRepository.delete(cartItem);
+        shoppingCart.removeCartItem(cartItem);

```

Comment:
I don't think you need both of these statements

if you have a proper mapping - `removeCartItem` should be enough